File: 14-TypeScript/Grey Multi-Tenant/packages/grey-php/src/Error/GreyError.php
<?php

declare(strict_types=1);

namespace Grey\Sdk\Error;

use Throwable;

/**
 * Represents a normalized error in the Grey SDK.
 */
final class GreyError
{
    public function __construct(
        private readonly ErrorCode $code,
        private readonly string $message,
        private readonly ?array $details = null,
    ) {}

    /**
     * Gets the error code.
     */
    public function getCode(): ErrorCode
    {
        return $this->code;
    }

    /**
     * Gets the error code as a string.
     */
    public function getCodeString(): string
    {
        return $this->code->value;
    }

    /**
     * Gets the error message.
     */
    public function getMessage(): string
    {
        return $this->message;
    }

    /**
     * Gets additional error details.
     */
    public function getDetails(): ?array
    {
        return $this->details;
    }

    /**
     * Creates an unauthorized error.
     */
    public static function unauthorized(?string $message = null): self
    {
        return new self(
            ErrorCode::UNAUTHORIZED,
            $message ?? 'Authentication required',
        );
    }

    /**
     * Creates a forbidden error.
     */
    public static function forbidden(?string $message = null): self
    {
        return new self(
            ErrorCode::FORBIDDEN,
            $message ?? 'Permission denied',
        );
    }

    /**
     * Creates a not found error.
     */
    public static function notFound(?string $message = null): self
    {
        return new self(
            ErrorCode::NOT_FOUND,
            $message ?? 'Resource not found',
        );
    }

    /**
     * Creates a validation error.
     */
    public static function validation(string $message, ?array $details = null): self
    {
        return new self(
            ErrorCode::VALIDATION_ERROR,
            $message,
            $details,
        );
    }

    /**
     * Creates a network error.
     */
    public static function network(?string $message = null): self
    {
        return new self(
            ErrorCode::NETWORK_ERROR,
            $message ?? 'Network error occurred',
        );
    }

    /**
     * Creates a timeout error.
     */
    public static function timeout(?string $message = null): self
    {
        return new self(
            ErrorCode::TIMEOUT,
            $message ?? 'Request timed out',
        );
    }

    /**
     * Creates a server error.
     */
    public static function server(?string $message = null): self
    {
        return new self(
            ErrorCode::SERVER_ERROR,
            $message ?? 'Server error occurred',
        );
    }

    /**
     * Creates an unknown error.
     */
    public static function unknown(?string $message = null): self
    {
        return new self(
            ErrorCode::UNKNOWN,
            $message ?? 'An unknown error occurred',
        );
    }

    /**
     * Creates a GreyError from an HTTP response.
     */
    public static function fromHttpResponse(int $status, ?array $body = null): self
    {
        $code = ErrorCode::fromHttpStatus($status);
        $message = $body['message'] ?? $body['error'] ?? "HTTP error $status";
        $details = $body['details'] ?? null;

        return new self($code, $message, $details);
    }

    /**
     * Creates a GreyError from an exception.
     */
    public static function fromException(Throwable $e): self
    {
        if ($e instanceof \GuzzleHttp\Exception\ConnectException) {
            return self::network($e->getMessage());
        }

        if ($e instanceof \GuzzleHttp\Exception\RequestException) {
            $response = $e->getResponse();
            if ($response !== null) {
                $status = $response->getStatusCode();
                $body = json_decode((string) $response->getBody(), true);
                return self::fromHttpResponse($status, is_array($body) ? $body : null);
            }
            return self::network($e->getMessage());
        }

        return self::unknown($e->getMessage());
    }

    /**
     * Converts the error to an array.
     */
    public function toArray(): array
    {
        return [
            'code' => $this->code->value,
            'message' => $this->message,
            'details' => $this->details,
        ];
    }

    /**
     * Converts the error to JSON.
     */
    public function toJson(): string
    {
        return json_encode($this->toArray(), JSON_THROW_ON_ERROR);
    }
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-ruby/lib/grey_sdk/error/grey_error.rb
# frozen_string_literal: true

require "json"

module GreySdk
  module Error
    # Represents a normalized error in the Grey SDK
    class GreyError
      attr_reader :code, :message, :details

      # @param code [Symbol] Error code
      # @param message [String] Error message
      # @param details [Hash, nil] Additional error details
      def initialize(code, message, details = nil)
        @code = ErrorCode.normalize(code)
        @message = message.to_s
        @details = details
      end

      # Creates an unauthorized error
      # @param message [String, nil] Custom message
      # @return [GreyError]
      def self.unauthorized(message = nil)
        new(ErrorCode::UNAUTHORIZED, message || "Authentication required")
      end

      # Creates a forbidden error
      # @param message [String, nil] Custom message
      # @return [GreyError]
      def self.forbidden(message = nil)
        new(ErrorCode::FORBIDDEN, message || "Permission denied")
      end

      # Creates a not found error
      # @param message [String, nil] Custom message
      # @return [GreyError]
      def self.not_found(message = nil)
        new(ErrorCode::NOT_FOUND, message || "Resource not found")
      end

      # Creates a validation error
      # @param message [String] Error message
      # @param details [Hash, nil] Validation details
      # @return [GreyError]
      def self.validation(message, details = nil)
        new(ErrorCode::VALIDATION_ERROR, message, details)
      end

      # Creates a network error
      # @param message [String, nil] Custom message
      # @return [GreyError]
      def self.network(message = nil)
        new(ErrorCode::NETWORK_ERROR, message || "Network error occurred")
      end

      # Creates a timeout error
      # @param message [String, nil] Custom message
      # @return [GreyError]
      def self.timeout(message = nil)
        new(ErrorCode::TIMEOUT, message || "Request timed out")
      end

      # Creates a server error
      # @param message [String, nil] Custom message
      # @return [GreyError]
      def self.server(message = nil)
        new(ErrorCode::SERVER_ERROR, message || "Server error occurred")
      end

      # Creates an unknown error
      # @param message [String, nil] Custom message
      # @return [GreyError]
      def self.unknown(message = nil)
        new(ErrorCode::UNKNOWN, message || "An unknown error occurred")
      end

      # Creates a GreyError from an HTTP response
      # @param status [Integer] HTTP status code
      # @param body [Hash, nil] Response body
      # @return [GreyError]
      def self.from_http_response(status, body = nil)
        code = ErrorCode.from_http_status(status)
        message = body&.dig("message") || body&.dig("error") || "HTTP error #{status}"
        details = body&.dig("details")

        new(code, message, details)
      end

      # Creates a GreyError from an exception
      # @param exception [Exception] The exception
      # @return [GreyError]
      def self.from_exception(exception)
        case exception
        when Errno::ECONNREFUSED, Errno::ECONNRESET, Errno::EHOSTUNREACH
          network(exception.message)
        when Net::OpenTimeout, Net::ReadTimeout
          timeout(exception.message)
        when JSON::ParserError
          validation("Invalid JSON response: #{exception.message}")
        else
          unknown(exception.message)
        end
      end

      # Converts the error to a hash
      # @return [Hash]
      def to_h
        {
          code: @code,
          message: @message,
          details: @details
        }
      end

      # Converts the error to JSON
      # @return [String]
      def to_json(*args)
        to_h.to_json(*args)
      end

      # String representation
      # @return [String]
      def to_s
        "GreyError[#{@code}]: #{@message}"
      end

      # Inspect representation
      # @return [String]
      def inspect
        "#<GreySdk::Error::GreyError code=#{@code.inspect} message=#{@message.inspect}>"
      end
    end
  end
end
